x-pack/plugins/enterprise_search/public/applications/app_search/components/analytics/analytics_logic.test.ts

```diff
@@ -0,0 +1,308 @@
+
+        expect(AnalyticsLogic.values).toEqual({
+          ...DEFAULT_VALUES,
+          dataLoading: false,
+          analyticsUnavailable: false,
+          // TODO: more state will get set here in future PRs
+        });
+      });
+    });
+  });
+
+  describe('listeners', () => {
+    describe('loadAnalyticsData', () => {
+      it('should set state', () => {
+        mount({ dataLoading: false });
+
+        AnalyticsLogic.actions.loadAnalyticsData();
+
+        expect(AnalyticsLogic.values).toEqual({
+          ...DEFAULT_VALUES,
+          dataLoading: true,
+        });
+      });
+
+      it('should make an API call and set state based on the response', async () => {
+        const promise = Promise.resolve(MOCK_ANALYTICS_RESPONSE);
+        (HttpLogic.values.http.get as jest.Mock).mockReturnValueOnce(promise);
+        mount();
+        jest.spyOn(AnalyticsLogic.actions, 'onAnalyticsDataLoad');
+
+        AnalyticsLogic.actions.loadAnalyticsData();
+        await promise;
+
+        expect(HttpLogic.values.http.get).toHaveBeenCalledWith(
+          '/api/app_search/engines/test-engine/analytics/queries',
+          {
+            query: { size: 20 },
+          }
+        );
+        expect(AnalyticsLogic.actions.onAnalyticsDataLoad).toHaveBeenCalledWith(
+          MOCK_ANALYTICS_RESPONSE
+        );
+      });
+
+      it('parses and passes the current search query string', async () => {
+        (HttpLogic.values.http.get as jest.Mock).mockReturnValueOnce({});
+        KibanaLogic.values.history.location.search =
+          '?start=1970-01-01&end=1970-01-02&&tag=some_tag';
+        mount();
+
+        AnalyticsLogic.actions.loadAnalyticsData();
+
+        expect(HttpLogic.values.http.get).toHaveBeenCalledWith(
+          '/api/app_search/engines/test-engine/analytics/queries',
+          {
+            query: {
+              start: '1970-01-01',
+              end: '1970-01-02',
+              tag: 'some_tag',
+              size: 20,
+            },
+          }
+        );
+      });
+
+      it('calls onAnalyticsUnavailable if analyticsUnavailable is in response', async () => {
+        const promise = Promise.resolve({ analyticsUnavailable: true });
+        (HttpLogic.values.http.get as jest.Mock).mockReturnValueOnce(promise);
+        mount();
+        jest.spyOn(AnalyticsLogic.actions, 'onAnalyticsUnavailable');
+
+        AnalyticsLogic.actions.loadAnalyticsData();
+        await promise;
+
+        expect(AnalyticsLogic.actions.onAnalyticsUnavailable).toHaveBeenCalled();
+      });
+
+      it('handles errors', async () => {
+        const promise = Promise.reject('error');
+        (HttpLogic.values.http.get as jest.Mock).mockReturnValueOnce(promise);
+        mount();
+        jest.spyOn(AnalyticsLogic.actions, 'onAnalyticsUnavailable');
+
+        try {
+          AnalyticsLogic.actions.loadAnalyticsData();
+          await promise;
+        } catch {
+          // Do nothing
+        }
```

Comment:
Do we expect `loadAnalyticsData` to throw an Error the http fails? It looks like [loadAnalyticsData](https://github.com/elastic/kibana/pull/88250/files#diff-16164984dc6aaccd02e30597a3f3669856a906b10b1c5d539e1d3714647cba47R76-R79) is setup to catch the any http errors.  Alternatively, I think you should explicitly assert that `loadAnalyticsData` is throwing.

Reply:
> Do we expect loadAnalyticsData to throw an Error the http fails

Yes. Rejected promises **as well as** 400/500 responses from the server (in production) will trigger the `catch` block.

> It looks like loadAnalyticsData is setup to catch the any http errors.

You can try this yourself if you pull the branch down, removing this try catch will cause the test to throw an error. I believe this actually might be because `flashAPIErrors()` re-throws the error if there isn't a valid message to display.

>  Alternatively, I think you should explicitly assert that loadAnalyticsData is throwing.

Unfortunately it's not quite that simple with Jest and async expects that are actually throws, the syntax is fairly bizarre and far more complex than this. I've struggled with this a lot in the past and am confident try/catch within the test is by far the easiest way to catch expected errors.

I definitely agree we could clarify the `// Do nothing` comment to be a little more helpful and indicate why we're doing nothing, although that would probably have to be a separate PR since I'm essentially copying a pattern we've done elsewhere in the codebase (originally set up by @JasonStoltz). We could also possibly extract this out to a helper which could help clear confusion in that regard.

Reply:
It's complicated, IIRC.

You can't catch an exception thrown by a listener, like `loadAnalyticsData`.

I think it's cause you're not invoking the listener directly. They're on separate call stacks or something. It's like an evented system. Calling `loadAnalyticsData` broadcasts an event, and then 1 or more event listeners are triggered and execute. They're separate run contexts / call stacks / or however you would say that in JS. So you could wrap your code that triggers an event in a try / catch, and it will catch errors that occur while triggering that event, but it won't catch errors that occur in event handlers.

Does that much make sense?

So this code is actually misleading. You could actually move `loadAnalyticsData` outside of the try / catch and it would work ( I didn't test this, so please tell me to stuff it if I'm wrong 😰 )

```
        AnalyticsLogic.actions.loadAnalyticsData();
        
        try {
          await promise;
        } catch {
          // Do nothing
        }
```

So the point so far is that this test isn't actually testing that loadAnalyticsData throws correctly. This test is testing that the `loadAnalyticsData` correctly handles a rejected promise.

So here's the next challenge.

We need to tell `http.get` to return a reject promise so that we can test that scenario. And then we need to `await` that promise.

You can't do something like this:

```
(HttpLogic.values.http.get as jest.Mock).mockReturnValueOnce(Promise.reject());
await AnalyticsLogic.actions.loadAnalyticsData();
```

Because `loadAnalyticsData` doesn't return a promise. So we have to create the promise separately so we can hold a reference to it to await later:

```
        const promise = Promise.reject('error');
        (HttpLogic.values.http.get as jest.Mock).mockReturnValueOnce(promise);
        await promise;
```

When you await a rejected promise, it will throw, so you gotta wrap it up:

```
        const promise = Promise.reject('error');
        (HttpLogic.values.http.get as jest.Mock).mockReturnValueOnce(promise);

        try {
            await promise; 
        } catch {
          // Do nothing
        }
```

You may be tempted to throw some assertions in the catch at this point..

```
        const promise = Promise.reject('error');
        (HttpLogic.values.http.get as jest.Mock).mockReturnValueOnce(promise);

        try {
            await promise; 
        } catch {
          expect(flashAPIErrors).toHaveBeenCalledWith('error');
          expect(AnalyticsLogic.actions.onAnalyticsUnavailable).toHaveBeenCalled();
        }
```

But I think you should generally avoid that ... because if for some reason your code ever stops throwing, then your code won't run the expectations, but the code will still pass. 

```
        const promise = Promise.reject('error');
        (HttpLogic.values.http.get as jest.Mock).mockReturnValueOnce(promise);

        try {
            await promise; 
        } catch {
            // Do nothing
        }
        
          expect(flashAPIErrors).toHaveBeenCalledWith('error');
          expect(AnalyticsLogic.actions.onAnalyticsUnavailable).toHaveBeenCalled();
```

So that's generally how we arrived here. This could probably be improved.

Any ideas?

Reply:
If you can get access to invoke a listener directly, that could simplify things.

Reply:
Follow up PR: https://github.com/elastic/kibana/pull/88422